agent_app_sample_code/02_data_pipeline.py

```diff
@@ -0,0 +1,421 @@
+# MAGIC
+# MAGIC After you test your POC with stakeholders, you can return here to change the parsing logic or extraction additional metadata about the documents to help improve the quality of your retriever.
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Define your parsing function
+# MAGIC
+# MAGIC This default implementation parses PDF, HTML, and DOCX files using open source libraries.  Adjust `file_parser(...)` to add change the parsing logic, add support for more file types, or extract additional metadata about each document.
+
+# COMMAND ----------
+
+from typing import TypedDict
+from datetime import datetime
+import warnings
+import io
+import traceback
+
+# PDF libraries
+from pypdf import PdfReader
+
+# HTML libraries
+from markdownify import markdownify as md
+import markdownify
+import re
+
+## DOCX libraries
+import pypandoc
+import tempfile
+
+# Schema of the dict returned by `file_parser(...)`
+class ParserReturnValue(TypedDict):
+    # DO NOT CHANGE THESE NAMES - these are required by Evaluation & Framework
+    # Parsed content of the document
+    doc_content: str  # do not change this name
+    # The status of whether the parser succeeds or fails, used to exclude failed files downstream
+    parser_status: str  # do not change this name
+    # Unique ID of the document
+    doc_uri: str  # do not change this name
+
+    # OK TO CHANGE THESE NAMES
+    # Optionally, you can add additional metadata fields here
+    example_metadata: str
+    last_modified: datetime
+
+
+# Parser function.  Replace this function to provide custom parsing logic.
+def file_parser(
+    raw_doc_contents_bytes: bytes,
+    doc_path: str,
+    modification_time: datetime,
+    doc_bytes_length: int,
+) -> ParserReturnValue:
+    """
+    Parses the content of a PDF document into a string.
+
+    This function takes the raw bytes of a PDF document and its path, attempts to parse the document using PyPDF,
+    and returns the parsed content and the status of the parsing operation.
+
+    Parameters:
+    - raw_doc_contents_bytes (bytes): The raw bytes of the document to be parsed (set by Spark when loading the file)
+    - doc_path (str): The path of the document, used to verify the file extension (set by Spark when loading the file)
+    - modification_time (timestamp): The last modification time of the document (set by Spark when loading the file)
+    - doc_bytes_length (long): The size of the document in bytes (set by Spark when loading the file)
+
+    Returns:
+    - ParserReturnValue: A dictionary containing the parsed document content and the status of the parsing operation.
+      The 'doc_content' key will contain the parsed text as a string, and the 'parser_status' key will indicate
+      whether the parsing was successful or if an error occurred.
+    """
+    try:
+        file_extension = doc_path.split(".")[-1]
```

Comment:
Should probably use official APIs to get the file extension e.g. https://stackoverflow.com/a/541394